File: easy/160. Intersection of Two Linked Lists/160. Intersection of Two Linked Lists.ts
/**
 * Definition for singly-linked list.
 */

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
    let first = headA, second = headB
    const ht = new Map()

    while (first) {
        ht.set(first, true)
        first = first.next
    }

    while (second) {
        if (ht.has(second)) return second
        second = second.next
    }

    return null
};

export { }

File: easy/160. Intersection of Two Linked Lists/160. Intersection of Two Linked Lists.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def getIntersectionNode(
        self, headA: Optional[ListNode], headB: Optional[ListNode]
    ) -> Optional[ListNode]:
        first, second = headA, headB
        ht = {}

        while first:
            ht[first] = True
            first = first.next

        while second:
            if ht[second]:
                return second
            second = second.next

        return None
